ocrd_eval.md

```diff
@@ -0,0 +1,372 @@
+
+## Evaluation Metrics
+
+The evaluation of the success (accuracy) of OCR is a complex task for which multiple methods and metrics are available. It aims to capture quality in different aspects, such as the recognition of text, but also the detection of layout, for which different methods and metrics are needed.
+
+Furthermore, the time and resources required for OCR processing also have to be captured. Here we describe the metrics that were selected for use in OCR-D, how exactly they are applied, and what was the motivation.
+
+### Scope of These Definitions
+
+At this stage (Q3 2022) these definitions serve as a basis of common understanding for the metrics used in the benchmarking presented in OCR-D QUIVER. Further implications for evaluation tools do not yet apply.
+
+### Text Evaluation
+
+The most important measure to assess the quality of OCR is the accuracy of the recognized text. The majority of metrics for this are based on the Levenshtein distance, an algorithm to compute the distance between two strings. In OCR, one of these strings is generally the Ground Truth text and the other the recognized text which is the result of an OCR.
+
+#### Levenshtein Distance (Edit Distance)
+
+Levenshtein distance between two strings `a` and `b` is the number of edit operations needed to turn `a` into `b`. Edit operations depend on the specific variant of the algorithm but for OCR, relevant operations are deletion, insertion and substitution.
+
+The Levenshtein distance forms the basis for the calculation of [CER/WER](https://pad.gwdg.de/#CERWER).
+As there are different implementations of the edit distance available (e.g. [rapidfuzz](https://maxbachmann.github.io/RapidFuzz/Usage/distance/Levenshtein.html), [jellyfish](https://jamesturk.github.io/jellyfish/functions/#levenshtein-distance), …), the OCR-D coordination project will provide a recommendation in the final version of this document.
+
+##### Example
+
+Given a Ground truth that reads `ſind` and the recognized text `fmd`.
+
+The Levenshtein distance between these texts is 4, because 4 edit operations are necessary to turn `fmd` into `ſind`:
+
+* `fmd` --> `ſmd` (substitution)
+* `ſmd` --> `ſimd` (insertion)
+* `ſimd` --> `ſind` (substitution)
+
+#### CER and WER
+
+##### Characters
+
+A text consists of a set of characters that have a certain meaning. A character is a glyph that represents a word, a letter in a word, or a symbol.
+Not included in the character definition are all forms of white spaces.
+
+###### Examples
+
+* the character `a` in the text `babst` represents the German letter `a`
+* the character `&` represents the Latin abbreviation `etc.`
+* the character `☿` represents an Astronomical symbol for the planet Mercury
+
+##### Character Error Rate (CER)
+
+The character error rate (CER) describes how many faulty characters the output of an OCR engine contains compaired to the Ground Truth text in relation to the text length (i.e. the number of characters of the text in the GT)
```

Comment:
That's same with all this error-like stuff :slightly_smiling_face: 
We have to be careful not to interweave CER and CA (cf. https://www.o-bib.de/bib/article/view/5888/8845) 

The difference between the number of GT-chars and the edit distance for GT and text (which is considered to reflect misses) represents the number of correct recognized characters. As long as this difference is greater than zero (read: there are less errors than total GT chars), one can go for the ratio of correct recognized chars vs. total GT chars.

Or, from an CER-perspective: As long as the edit distance (=number of incorrect recognized chars) is less than the number of GT-chars, go for the ratio.

Otherwise, CER is 100% (or 0% for CA).

Reply:
@cneud @kba I'd like to hear your opinion regarding this topic. What do you think?

Reply:
Like I said, this is not the right place to discuss the denominator (see below). My suggestion is **neutral** in that respect. It also avoids the term `faulty` and properly introduces error rate in relation to edit distance defined earlier.

Reply:
As for the question about what the denominator is, cf. https://github.com/impactcentre/ocrevalUAtion/issues/21

I think chosing the length of the edit distance path is a sensible approach because it avoids the `> 100%` problem.

Reply:
@bertsky @M3ssman 

@kba, @cneud and I had a discussion yesterday and we suggest that we go for

`length of edit distance path + correctly recognized characters`

as denominator as described here under "normalized CER". Describing the non-normalized CER where the denominator equals the number of chars in the GT is not necessary and confused us as well since we actually wanted to go for the **normalized** CER all along.

That being said, I'll also add a paragraph about this not being the final word in the overall discussion about CER in the OCR community. We choose what's practical for our case but are aware that this is a highly debated topic.

Reply:
@mweidling your statement got me confused.

> length of edit distance path + correctly recognized characters

Let me unpack this:
- There is no `edit distance path`; there is an `alignment path` for the alignment calculated to determine the edit distance. That alignment path length is by definition equal to $i+s+d+c$.
- That already is denominator (in the normalized case), so why the `+`?

Also, again, if you look at my suggestion above, in this paragraph there is **no need** to commit to normalized or unnormalized. My formulation is neutral (just `length`). (The place to make that commitment is below, after you introduced normalized and unnormalized.) Please consider applying the suggestion here.

Reply:
I meant alignment path. :) Thanks for clarifying!

And I understand that you don't say "use this or that". But we (as in Coordination Project) want to commit to one definition of the CER to have a sound base for calculating it in QuiVer. That's why we had the commitment in the draft from early on.

Reply:
> But we (as in Coordination Project) want to commit to one definition of the CER to have a sound base for calculating it in QuiVer. That's why we had the commitment in the draft from early on.

Look, my concern is with readibility here. In the paragraph we are talking about, it is simply too early to make that distinction and be that precise. You already have the details below. That's where you can formulate that commitment. 

Also, you current formulation still favours GT length here. My point is: why not simply apply my suggestion?